File: backend/fleets_manager.py
from os import path, getcwd

from aws_cdk import (
    Stack,
    Duration,
    DockerImage,
    aws_appsync_alpha as appsync,
    aws_dynamodb as dynamodb,
    aws_lambda_python_alpha as lambda_python,
    aws_lambda as awslambda,
    aws_cognito as cognito,
    aws_iam as iam,
)

from constructs import Construct


class FleetsManager(Construct):
    def __init__(
        self,
        scope: Construct,
        id: str,
        api: appsync.IGraphqlApi,
        user_pool: cognito.IUserPool,
        roles_to_grant_invoke_access: list[iam.IRole],
        **kwargs,
    ):
        super().__init__(scope, id, **kwargs)

        stack = Stack.of(self)

        fleets_table = dynamodb.Table(
            self,
            "FleetsTable",
            partition_key=dynamodb.Attribute(
                name="fleetId", type=dynamodb.AttributeType.STRING
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            encryption=dynamodb.TableEncryption.AWS_MANAGED,
        )

        lambda_architecture = awslambda.Architecture.ARM_64
        lambda_runtime = awslambda.Runtime.PYTHON_3_9
        lambda_bundling_image = DockerImage.from_registry(
            "public.ecr.aws/sam/build-python3.9:latest-arm64"
        )

        fleets_handler = lambda_python.PythonFunction(
            self,
            "fleetsFunction",
            entry="backend/lambdas/fleets_function/",
            description="Fleets Resolver",
            index="index.py",
            handler="lambda_handler",
            timeout=Duration.minutes(1),
            runtime=lambda_runtime,
            tracing=awslambda.Tracing.ACTIVE,
            memory_size=128,
            architecture=lambda_architecture,
            bundling=lambda_python.BundlingOptions(image=lambda_bundling_image),
            # layers=[helper_functions_layer, powertools_layer]
            layers=[
                lambda_python.PythonLayerVersion.from_layer_version_arn(
                    self,
                    "powertools",
                    "arn:aws:lambda:eu-west-1:017000801446:layer:AWSLambdaPowertoolsPython:3",
                )
            ],
            environment={
                "DDB_TABLE": fleets_table.table_name,
                "user_pool_id": user_pool.user_pool_id,
            },
        )

        fleets_table.grant_read_write_data(fleets_handler)

        # Define the data source for the API
        fleets_data_source = api.add_lambda_data_source(
            "FleetsDataSource", fleets_handler
        )

        none_data_source = api.add_none_data_source("fleets_none")
        # Define API Schema

        fleets_object_Type = appsync.ObjectType(
            "Fleet",
            definition={
                "fleetName": appsync.GraphqlType.string(),
                "fleetId": appsync.GraphqlType.id(),
                "createdAt": appsync.GraphqlType.aws_date_time(),
            },
        )

        api.add_type(fleets_object_Type)

        # Fleet methods
        api.add_query(
            "getAllFleets",
            appsync.ResolvableField(
                return_type=fleets_object_Type.attribute(is_list=True),
                data_source=fleets_data_source,
            ),
        )
        api.add_mutation(
            "addFleet",
            appsync.ResolvableField(
                args={
                    "fleetName": appsync.GraphqlType.string(is_required=True),
                },
                return_type=fleets_object_Type.attribute(),
                data_source=fleets_data_source,
            ),
        )
        api.add_subscription(
            "addedFleet",
            appsync.ResolvableField(
                return_type=fleets_object_Type.attribute(),
                data_source=none_data_source,
                request_mapping_template=appsync.MappingTemplate.from_string(
                    """{
                        "version": "2017-02-28",
                        "payload": $util.toJson($context.arguments.entry)
                    }"""
                ),
                response_mapping_template=appsync.MappingTemplate.from_string(
                    """$util.toJson($context.result)"""
                ),
                directives=[appsync.Directive.subscribe("addFleet")],
            ),
        )

        api.add_mutation(
            "deleteFleet",
            appsync.ResolvableField(
                args={"fleetId": appsync.GraphqlType.string(is_required=True)},
                return_type=fleets_object_Type.attribute(),
                data_source=fleets_data_source,
            ),
        )
        api.add_subscription(
            "deletedFleet",
            appsync.ResolvableField(
                return_type=fleets_object_Type.attribute(),
                data_source=none_data_source,
                request_mapping_template=appsync.MappingTemplate.from_string(
                    """{
                        "version": "2017-02-28",
                        "payload": $util.toJson($context.arguments.entry)
                    }"""
                ),
                response_mapping_template=appsync.MappingTemplate.from_string(
                    """$util.toJson($context.result)"""
                ),
                directives=[appsync.Directive.subscribe("deleteFleet")],
            ),
        )

        api.add_mutation(
            "updateFleet",
            appsync.ResolvableField(
                args={
                    "fleetId": appsync.GraphqlType.string(is_required=True),
                    "fleetName": appsync.GraphqlType.string(),
                },
                return_type=fleets_object_Type.attribute(),
                data_source=fleets_data_source,
            ),
        )
        api.add_subscription(
            "updatedFleet",
            appsync.ResolvableField(
                return_type=fleets_object_Type.attribute(),
                data_source=none_data_source,
                request_mapping_template=appsync.MappingTemplate.from_string(
                    """{
                        "version": "2017-02-28",
                        "payload": $util.toJson($context.arguments.entry)
                    }"""
                ),
                response_mapping_template=appsync.MappingTemplate.from_string(
                    """$util.toJson($context.result)"""
                ),
                directives=[appsync.Directive.subscribe("updateFleet")],
            ),
        )

        # Grant access so API methods can be invoked
        for role in roles_to_grant_invoke_access:
            role.add_to_policy(
                iam.PolicyStatement(
                    effect=iam.Effect.ALLOW,
                    actions=["appsync:GraphQL"],
                    resources=[
                        f"{api.arn}/types/Query/fields/getAllFleets",
                        f"{api.arn}/types/Mutation/fields/addFleet",
                        f"{api.arn}/types/Subscription/fields/addedFleet",
                        f"{api.arn}/types/Mutation/fields/deleteFleet",
                        f"{api.arn}/types/Subscription/fields/deletedFleet",
                        f"{api.arn}/types/Mutation/fields/updateFleet",
                        f"{api.arn}/types/Subscription/fields/addedFleet",
                        f"{api.arn}/types/Subscription/fields/deletedFleet",
                        f"{api.arn}/types/Subscription/fields/updatedFleet",
                    ],
                )
            )

File: lib/constructs/fleets-manager.ts
import { DockerImage, Duration, RemovalPolicy } from 'aws-cdk-lib';
import * as appsync from 'aws-cdk-lib/aws-appsync';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import { EventBus } from 'aws-cdk-lib/aws-events';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { CodeFirstSchema, Directive, GraphqlType, ObjectType, ResolvableField } from 'awscdk-appsync-utils';
import { NagSuppressions } from 'cdk-nag';
import { StandardLambdaPythonFunction } from './standard-lambda-python-function';

import { Construct } from 'constructs';

export interface FleetsManagerProps {
  userPoolId: string;
  appsyncApi: {
    schema: CodeFirstSchema;
    api: appsync.GraphqlApi;
    noneDataSource: appsync.NoneDataSource;
  };
  lambdaConfig: {
    runtime: lambda.Runtime;
    architecture: lambda.Architecture;
    bundlingImage: DockerImage;
    layersConfig: {
      powerToolsLogLevel: string;
      helperFunctionsLayer: lambda.ILayerVersion;
      powerToolsLayer: lambda.ILayerVersion;
      appsyncHelpersLayer: lambda.ILayerVersion;
    };
  };
  eventbus: EventBus;
}

export class FleetsManager extends Construct {
  constructor(scope: Construct, id: string, props: FleetsManagerProps) {
    super(scope, id);

    const fleets_table = new dynamodb.Table(this, 'FleetsTable', {
      partitionKey: {
        name: 'fleetId',
        type: dynamodb.AttributeType.STRING,
      },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      encryption: dynamodb.TableEncryption.AWS_MANAGED,
      removalPolicy: RemovalPolicy.DESTROY,
      pointInTimeRecovery: true,
    });

    const fleets_handler = new StandardLambdaPythonFunction(this, 'fleetsFunction', {
      entry: 'lib/lambdas/fleets_function/',
      description: 'Fleets Resolver',
      index: 'index.py',
      handler: 'lambda_handler',
      timeout: Duration.minutes(1),
      runtime: props.lambdaConfig.runtime,
      memorySize: 128,
      architecture: props.lambdaConfig.architecture,
      bundling: {
        image: props.lambdaConfig.bundlingImage,
      },
      layers: [props.lambdaConfig.layersConfig.powerToolsLayer, props.lambdaConfig.layersConfig.appsyncHelpersLayer],
      environment: {
        DDB_TABLE: fleets_table.tableName,
        user_pool_id: props.userPoolId,
        POWERTOOLS_SERVICE_NAME: 'fleets_resolver',
        LOG_LEVEL: props.lambdaConfig.layersConfig.powerToolsLogLevel,
        APPSYNC_URL: props.appsyncApi.api.graphqlUrl,
      },
    });

    fleets_table.grantReadWriteData(fleets_handler);
    props.appsyncApi.api.grantMutation(fleets_handler, 'carsUpdateFleet');

    // Define the data source for the API
    const fleets_data_source = props.appsyncApi.api.addLambdaDataSource('FleetsDataSource', fleets_handler);

    NagSuppressions.addResourceSuppressions(
      fleets_data_source,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Suppress wildcard that covers Lambda aliases in resource path',
          appliesTo: [
            {
              regex: '/^Resource::(.+):\\*$/g',
            },
          ],
        },
      ],
      true
    );

    // Define API Schema
    const fleets_object_Type = new ObjectType('Fleet', {
      definition: {
        fleetName: GraphqlType.string(),
        fleetId: GraphqlType.id(),
        createdAt: GraphqlType.awsDateTime(),
        createdBy: GraphqlType.id(),
        carIds: GraphqlType.id({ isList: true }),
      },
      directives: [Directive.cognito('admin', 'commentator', 'operator')],
    });

    props.appsyncApi.schema.addType(fleets_object_Type);

    // Fleet methods
    props.appsyncApi.schema.addQuery(
      'getAllFleets',
      new ResolvableField({
        returnType: fleets_object_Type.attribute({ isList: true }),
        dataSource: fleets_data_source,
        directives: [Directive.cognito('admin', 'operator')],
      })
    );

    props.appsyncApi.schema.addMutation(
      'addFleet',
      new ResolvableField({
        args: {
          fleetName: GraphqlType.string({ isRequired: true }),
          carIds: GraphqlType.string({ isList: true }),
        },
        returnType: fleets_object_Type.attribute(),
        dataSource: fleets_data_source,
        directives: [Directive.cognito('admin', 'operator')],
      })
    );
    props.appsyncApi.schema.addSubscription(
      'onAddedFleet',
      new ResolvableField({
        returnType: fleets_object_Type.attribute(),
        dataSource: props.appsyncApi.noneDataSource,
        requestMappingTemplate: appsync.MappingTemplate.fromString(
          `{
                        "version": "2017-02-28",
                        "payload": $util.toJson($context.arguments.entry)
                    }`
        ),
        responseMappingTemplate: appsync.MappingTemplate.fromString('$util.toJson($context.result)'),
        directives: [Directive.subscribe('addFleet'), Directive.cognito('admin', 'operator')],
      })
    );

    props.appsyncApi.schema.addMutation(
      'deleteFleets',
      new ResolvableField({
        args: { fleetIds: GraphqlType.string({ isRequiredList: true }) },
        returnType: fleets_object_Type.attribute({ isList: true }),
        dataSource: fleets_data_source,
      })
    );
    props.appsyncApi.schema.addSubscription(
      'onDeletedFleets',
      new ResolvableField({
        returnType: fleets_object_Type.attribute({ isList: true }),
        dataSource: props.appsyncApi.noneDataSource,
        requestMappingTemplate: appsync.MappingTemplate.fromString(
          `{
                        "version": "2017-02-28",
                    "payload": $util.toJson($context.arguments.entry)
                    }`
        ),
        responseMappingTemplate: appsync.MappingTemplate.fromString('$util.toJson($context.result)'),
        directives: [Directive.subscribe('deleteFleets'), Directive.cognito('admin', 'operator')],
      })
    );

    props.appsyncApi.schema.addMutation(
      'updateFleet',
      new ResolvableField({
        args: {
          fleetId: GraphqlType.string({ isRequired: true }),
          fleetName: GraphqlType.string(),
          carIds: GraphqlType.id({ isList: true }),
        },
        returnType: fleets_object_Type.attribute(),
        dataSource: fleets_data_source,
        directives: [Directive.cognito('admin', 'operator')],
      })
    );
    props.appsyncApi.schema.addSubscription(
      'onUpdatedFleet',
      new ResolvableField({
        returnType: fleets_object_Type.attribute(),
        dataSource: props.appsyncApi.noneDataSource,
        requestMappingTemplate: appsync.MappingTemplate.fromString(
          `{
                        "version": "2017-02-28",
                    "payload": $util.toJson($context.arguments.entry)
                    }`
        ),
        responseMappingTemplate: appsync.MappingTemplate.fromString('$util.toJson($context.result)'),
        directives: [Directive.subscribe('updateFleet'), Directive.cognito('admin', 'operator')],
      })
    );
  }
}
